bot/exts/summer_aoc.py

```diff
@@ -0,0 +1,319 @@
+
+        if self.is_running:
+            if self.is_configured():
+                await self.start_event()
+            else:
+                log.error("Summer AoC state incomplete, failed to start event")
+                self.is_running = False
+                self.save_event_state()
+
+    async def save_event_state(self) -> None:
+        """Save the current state in redis."""
+        await self.cache.update({
+            "is_running": self.is_running,
+            "year": self.year,
+            "current_day": self.current_day,
+            "day_interval": self.day_interval,
+            "post_time": self.post_time,
+        })
+
+    async def start_event(self) -> None:
+        """Start event by recording state and creating async tasks to post puzzles."""
+        log.info(f"Starting Summer AoC event with {self.year=} {self.current_day=} {self.day_interval=}")
+        sleep_for = time_until(hour=self.post_time)
+        self.wait_task = asyncio.create_task(asyncio.sleep(sleep_for.total_seconds()))
+        log.debug(f"Waiting until {self.post_time}:00 UTC to start Summer AoC loop")
+        await self.wait_task
+
+        self.loop_task = tasks.Loop(
+            self.post_puzzle,
+            seconds=0,
+            minutes=0,
+            hours=(self.day_interval * 24),
+            time=MISSING,
+            count=None,
+            reconnect=True,
+        )
+        self.loop_task.start()
+
+    async def stop_event(self) -> bool:
+        """Cancel any active Summer AoC tasks. Returns whether the event was running."""
+        was_waiting = self.wait_task and not self.wait_task.done()
+        was_looping = self.loop_task and self.loop_task.is_running()
+        if was_waiting and was_looping:
+            log.error("Both wait and loop tasks were active. Both should now be cancelled.")
+
+        if was_waiting:
+            self.wait_task.cancel()
+            log.debug("Summer AoC stopped during wait task")
+
+        if was_looping:
+            self.loop_task.cancel()  # .cancel() doesn't allow the current iteration to finish
+            log.debug("Summer AoC stopped during loop task")
+
+        self.is_running = False
+        await self.save_event_state()
+        log.info("Summer AoC event stopped")
+        return was_waiting or was_looping
+
+    async def post_puzzle(self) -> None:
+        """Create a thread for the current day's puzzle."""
+        if self.current_day > LAST_DAY:
+            log.error("Attempted to post puzzle after last day, stopping event")
+            await self.stop_event()
+            return
+
+        log.info(f"Posting puzzle for day {self.current_day}")
+        channel: discord.TextChannel = self.bot.get_channel(Channels.summer_aoc_main)
+        thread_starter = await channel.send(
+            f"<@&{Roles.summer_aoc}>",
+            embed=self.get_puzzle_embed(),
+        )
+        await thread_starter.create_thread(name=f"Day {self.current_day} Spoilers")
```

Comment:
Actually rather than putting the year in the thread title, how about in the post's hyperlink?

<img width="447" alt="image" src="https://user-images.githubusercontent.com/75038675/182284763-a1f44e69-873b-46c2-b86b-281ef83c0a66.png">

Reply:
I thought it was more confusing than helpful, and since they have to go to AoC website anyway it'd be kind of redundant.